
from dataclasses import dataclass
import typing
import typing_extensions

from flytekit.core.annotation import FlyteAnnotation

from latch.types.metadata import NextflowParameter
from latch.types.file import LatchFile
from latch.types.directory import LatchDir, LatchOutputDir

# Import these into your `__init__.py` file:
#
# from .parameters import generated_parameters

generated_parameters = {
    'input': NextflowParameter(
        type=LatchFile,
        default=None,
        section_title='Input/output options',
        description='Path to comma-separated file containing information about the samples in the experiment.',
    ),
    'outdir': NextflowParameter(
        type=typing.Optional[typing_extensions.Annotated[LatchDir, FlyteAnnotation({'output': True})]],
        default=None,
        section_title=None,
        description='Path to the output directory where the results will be saved.',
    ),
    'email': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Email address for completion summary.',
    ),
    'multiqc_title': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='MultiQC report title. Printed as page header, used for filename if not otherwise specified.',
    ),
    'spot_intervals': NextflowParameter(
        type=typing.Optional[LatchFile],
        default=None,
        section_title='SSDS options',
        description='SSDS enrichment can be tested in defined genomic intervals by the pipeline. Intervals are specified in a config file.',
    ),
    'spot_intervals_genome': NextflowParameter(
        type=typing.Optional[str],
        default='null',
        section_title=None,
        description='Index to the genome in --spot_intervals config file. If not provided, the pipeline will attempt to use the genome (if defined).',
    ),
    'spot_interval_beds': NextflowParameter(
        type=typing.Optional[LatchFile],
        default=None,
        section_title=None,
        description='Should not be provided manually; autogenerated from --spot_intervals.',
    ),
    'parse_extra_types': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description="SSDS identifies ssDNA using the micro-homology structure at the 5' end of reads. When set to true, the SSDS pipeline also infers three additional, but low confidence fragment types; importantly, reads of these thress types are a best guess, but could be incorrect (i.e. fragments classified as dsDNA may instead come from ssDNA in a sufficiently enriched library).",
    ),
    'genome': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title='Reference genome options',
        description='Name of iGenomes reference.',
    ),
    'fasta': NextflowParameter(
        type=typing.Optional[LatchFile],
        default=None,
        section_title=None,
        description='Path to FASTA genome file.',
    ),
    'bwa': NextflowParameter(
        type=typing.Optional[LatchFile],
        default=None,
        section_title=None,
        description='Path to the BWA index folder.',
    ),
}

